### PR TITLE
chore: Eradicate C-style casts and further clean up some code

### PR DESCRIPTION
**Pull request by jadebenn**
Kills pretty much all the C-style casts in our code base. Marking as draft until I have time to finish playtesting all the features this PR touches.

### CODE REVIEW

File: dCommon/Diagnostics.cpp [hunk 1]
```diff
@@ -197,10 +197,10 @@ void MakeBacktrace() {
 	sigact.sa_sigaction = CritErrHdlr;
 	sigact.sa_flags = SA_RESTART | SA_SIGINFO;
 
-	if (sigaction(SIGSEGV, &sigact, (struct sigaction*)nullptr) != 0 ||
-		sigaction(SIGFPE, &sigact, (struct sigaction*)nullptr) != 0 ||
-		sigaction(SIGABRT, &sigact, (struct sigaction*)nullptr) != 0 ||
-		sigaction(SIGILL, &sigact, (struct sigaction*)nullptr) != 0) {
+	if (sigaction(SIGSEGV, &sigact, static_cast<struct sigaction*>(nullptr)) != 0 ||
```
Comment (EmosewaMC) [inline dCommon/Diagnostics.cpp:200]: These shouldn't even be necessary

File: dCommon/NiPoint3.cpp [hunk 2]
```diff
@@ -232,10 +232,21 @@ NiPoint3 NiPoint3::MoveTowards(const NiPoint3& current, const NiPoint3& target,
...
+
+	if (static_cast<double>(lengthSquared) == 0.0
+		|| static_cast<double>(maxDistanceDelta) >= 0.0
+		&& static_cast<double>(lengthSquared)
+		<= static_cast<double>(maxDistanceDelta) * static_cast<double>(maxDistanceDelta)) {
 		return target;
-	float length = (float)std::sqrt((double)lengthSquared);
+	}
+
+	float length = static_cast<float>(std::sqrt(static_cast<double>(lengthSquared)));
```
Comment (EmosewaMC) [inline dCommon/NiPoint3.cpp:249]: Sqrtf will remove need for outer cast

File: dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp [hunk 3]
```diff
@@ -25,7 +25,7 @@ void CDMissionEmailTable::LoadValuesFromDatabase() {
 		entry.notificationGroup = tableData.getIntField("notificationGroup", -1);
 		entry.missionID = tableData.getIntField("missionID", -1);
 		entry.attachmentLOT = tableData.getIntField("attachmentLOT", 0);
-		entry.localize = (bool)tableData.getIntField("localize", -1);
+		entry.localize = static_cast<bool>(tableData.getIntField("localize", -1));
```
Comment (EmosewaMC) [inline dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp:28]: getBoolField

File: dGame/dComponents/BaseCombatAIComponent.h [hunk 4]
```diff
@@ -19,7 +19,7 @@ class Entity;
 /**
  * The current state of the AI
  */
-enum class AiState : int {
+enum class AiState : uint32_t {
```
Comment (EmosewaMC) [inline dGame/dComponents/BaseCombatAIComponent.h:22]: Keep this as an int

Comment (jadebenn) [reply dGame/dComponents/BaseCombatAIComponent.h:22]: Why? Isn't the state serialized as a int32_t anyway?

File: dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp (hunk 3 above)
Comment (jadebenn) [reply dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp]: To my surprise, there is no getBoolField method

File: dGame/dComponents/BaseCombatAIComponent.h (hunk 4 above)
Comment (EmosewaMC) [reply dGame/dComponents/BaseCombatAIComponent.h:22]: Yeah I get what you mean now, feel free to add back in this change.

File: dCommon/Diagnostics.cpp [hunk 5]
```diff
@@ -197,17 +197,6 @@ void MakeBacktrace() {
...
-		sigaction(SIGFPE, &sigact, (struct sigaction*)nullptr) != 0 ||
-		sigaction(SIGABRT, &sigact, (struct sigaction*)nullptr) != 0 ||
-		sigaction(SIGILL, &sigact, (struct sigaction*)nullptr) != 0) {
-		fprintf(stderr, "error setting signal handler for %d (%s)\n",
-			SIGSEGV,
-			strsignal(SIGSEGV));
-
-		exit(EXIT_FAILURE);
-	}
-
```
Comment (EmosewaMC) [inline dCommon/Diagnostics.cpp:210]: Why is this removed?

File: dCommon/NiPoint3.cpp [hunk 6]
```diff
@@ -114,13 +114,13 @@ bool NiPoint3::operator!=(const NiPoint3& point) const {
 //! Operator for subscripting
 float& NiPoint3::operator[](int i) {
 	float* base = &x;
-	return (float&)base[i];
+	return const_cast<float&>(base[i]);
```
Comment (EmosewaMC) [inline dCommon/NiPoint3.cpp:117]: This const cast shouldn't be needed

File: dCommon/SHA512.cpp [hunk 7]
```diff
@@ -51,7 +51,7 @@ void SHA512::transform(const unsigned char* message, unsigned int block_nb) {
 	uint64 t1, t2;
 	const unsigned char* sub_block;
 	int i, j;
-	for (i = 0; i < (int)block_nb; i++) {
+	for (i = 0; i < static_cast<int>(block_nb); i++) {
```
Comment (EmosewaMC) [inline dCommon/SHA512.cpp:54]: the changes in this file should be undone

File: dGame/dComponents/CharacterComponent.cpp [hunk 8]
```diff
@@ -157,7 +157,7 @@ void CharacterComponent::Serialize(RakNet::BitStream* outBitStream, bool bIsInit
 	outBitStream->Write(m_DirtySocialInfo);
 	if (m_DirtySocialInfo) {
 		outBitStream->Write(m_GuildID);
-		outBitStream->Write<unsigned char>(static_cast<unsigned char>(m_GuildName.size()));
+		outBitStream->Write<unsigned char>(m_GuildName.size());
 		if (!m_GuildName.empty())
 			outBitStream->WriteBits(reinterpret_cast<const unsigned char*>(m_GuildName.c_str()), static_cast<unsigned char>(m_GuildName.size()) * sizeof(wchar_t) * 8);
```
Comment (EmosewaMC) [inline dGame/dComponents/CharacterComponent.cpp:162]: This is just a reminder to myself (and maybe an easy issue for you to fix) that this line, while not touched in this PR, is incorrect and will cause issues on platforms where wchar_t is 4 bytes wide instead of 2 as it is on windows.

File: dGame/dGameMessages/GameMessages.cpp [hunk 9]
```diff
@@ -1950,7 +1950,7 @@ void GameMessages::SendBBBSaveResponse(const LWOOBJID& objectId, const LWOOBJID&
 		bitStream.Write(buffer[i]);
 
 	SEND_PACKET;
-	PacketUtils::SavePacket("eGameMessageType::BBB_SAVE_RESPONSE.bin", (char*)bitStream.GetData(), bitStream.GetNumberOfBytesUsed());
+	PacketUtils::SavePacket("eGameMessageType::BBB_SAVE_RESPONSE.bin", reinterpret_cast<char*>(bitStream.GetData()), bitStream.GetNumberOfBytesUsed());
```
Comment (EmosewaMC) [inline dGame/dGameMessages/GameMessages.cpp:1953]: This can honestly be removed/commented out as well

File: dScripts/02_server/Map/NS/Waves/ZoneNsWaves.cpp [hunk 10]
```diff
@@ -437,7 +437,7 @@ std::vector<Wave> ZoneNsWaves::GetWaves() {
 																									}, {}, {},
 																									"Treasure_Camera",
 																									5.0f,
-																									(uint32_t)-1,
+																									static_cast<uint32_t>(-1),
```
Comment (EmosewaMC) [inline dScripts/02_server/Map/NS/Waves/ZoneNsWaves.cpp:440]: yeah @aronwk-aaron i can see how this looks silly lol

File: dScripts/ai/MINIGAME/SG_GF/SERVER/SGCannon.cpp [hunk 11]
```diff
@@ -673,7 +673,7 @@ void SGCannon::RegisterHit(Entity* self, Entity* target, const std::string& time
 		target->GetPosition()
 	);
 
-	auto newScore = (int)self->GetVar<uint32_t>(TotalScoreVariable) + score;
+	auto newScore = self->GetVar<int32_t>(TotalScoreVariable) + score;
```
Comment (EmosewaMC) [inline dScripts/ai/MINIGAME/SG_GF/SERVER/SGCannon.cpp:676]: This actually must stay as uint32_t.  Variables on scripts are attached by not just key, but also T type.

**Review by EmosewaMC — CHANGES_REQUESTED** [review]
Great stuff, mostly on board, though there is some feedback that must be addressed, and some optionally addressed

File: dCommon/Diagnostics.cpp (hunk 5 above)
Comment (jadebenn) [reply dCommon/Diagnostics.cpp:210]: Didn't you tell me to remove it?

Comment (jadebenn) [reply dCommon/Diagnostics.cpp:210]: https://github.com/DarkflameUniverse/DarkflameServer/pull/1361#discussion_r1436625455

File: dCommon/SHA512.cpp (hunk 7 above)
Comment (jadebenn) [reply dCommon/SHA512.cpp:54]: Are we certain it is vendored? The closest I can find from a brief search is [this one,](https://www.geeksforgeeks.org/implement-secure-hashing-algorithm-512-sha-512-as-functional-programming-paradigm/) and even that's quite different in implementation.

File: dScripts/02_server/Map/NS/Waves/ZoneNsWaves.cpp (hunk 10 above)
Comment (jadebenn) [reply dScripts/02_server/Map/NS/Waves/ZoneNsWaves.cpp:440]: Any change I should make?

File: dScripts/ai/MINIGAME/SG_GF/SERVER/SGCannon.cpp (hunk 11 above)
Comment (jadebenn) [reply dScripts/ai/MINIGAME/SG_GF/SERVER/SGCannon.cpp:676]: I tried doing that, but it enables a roll-under bug if you shoot a friendly ship.

File: dCommon/Diagnostics.cpp (hunk 5 above)
Comment (EmosewaMC) [reply dCommon/Diagnostics.cpp:210]: Oh I meant the casts.  It's nullptr, it will always be cast to the correct ptr type

File: dScripts/02_server/Map/NS/Waves/ZoneNsWaves.cpp (hunk 10 above)
Comment (EmosewaMC) [reply dScripts/02_server/Map/NS/Waves/ZoneNsWaves.cpp:440]: Nah, just an observation as to why casting unsigned to signed looks weird

File: dScripts/ai/MINIGAME/SG_GF/SERVER/SGCannon.cpp (hunk 11 above)
Comment (EmosewaMC) [reply dScripts/ai/MINIGAME/SG_GF/SERVER/SGCannon.cpp:676]: Change all of the total score to be stored as int in that case.

**Review by EmosewaMC — COMMENTED** [review]
Has this been tested to progress missions and not underfloor on shooting gallery?

**Comment by jadebenn** [comment]
Tested both and they seem to work fine.

**Review by aronwk-aaron — APPROVED** [review]

File: dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp (hunk 3 above)
Comment (EmosewaMC) [inline dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp:28]: hm, an intended bool defaulting to -1 is kind of odd, this should probably just default to 0 instead.  Thoughts?

Comment (aronwk-aaron) [reply dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp:28]: default should be true (so 1)

**Review by aronwk-aaron — APPROVED** [review]

File: dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp (hunk 3 above)
Comment (jadebenn) [reply dDatabase/CDClientDatabase/CDClientTables/CDMissionEmailTable.cpp]: changed

**Review by EmosewaMC — APPROVED** [review]
